### PR TITLE
add sysinfo command

### PR DESCRIPTION
**Pull request by kovrus**

### CODE REVIEW

**Comment by chaudum** [comment]
requires version check!

File: CHANGES.txt [hunk 1]
```diff
@@ -5,6 +5,9 @@ Changes for crash
 Unreleased
 ==========
 
+ - Added the ``sysinfo`` command which queries the ``sys`` tables for
+   the system and cluste information.
```
Comment (chaudum) [inline CHANGES.txt]: typo: ~~the~~ system and **cluster** information

File: docs/cli.txt [hunk 2]
```diff
@@ -23,7 +23,8 @@ The ``crash`` binary supports several command line arguments.
 +---------------------------------------+--------------------------------------------------+
 | ``--version``                         | Print the crash version and exit.                |
 +---------------------------------------+--------------------------------------------------+
-
+| ``--sysinfo``                         | Print the system and cluster information.        |
```
Comment (chaudum) [inline docs/cli.txt]: ~~the~~

File: docs/cli.txt [hunk 3]
```diff
@@ -70,7 +71,8 @@ You can get a list of commands by typing the command ``\?``:
 +----------------------+------------------------------------------------------------------------------------+
 | ``\check``           | Query the sys.checks table for failing cluster checks                              |
 +----------------------+------------------------------------------------------------------------------------+
-
+| ``\sysinfo``         | Query the ``sys`` tables for the system and cluster information                    |
```
Comment (chaudum) [inline docs/cli.txt]: ~~the~~

File: src/crate/crash/command.py [hunk 4]
```diff
@@ -91,8 +92,13 @@ def parse_args(output_formats):
     parser.add_argument('--history',
                         type=str,
                         help='the history file to use', default=HISTORY_PATH)
-    parser.add_argument('-c', '--command', type=str,
+
+    group = parser.add_mutually_exclusive_group()
+    group.add_argument('-c', '--command', type=str,
                         help='execute sql statement')
+    group.add_argument('--sysinfo', action='store_true', default=False,
+                        help='show systmem information')
```
Comment (chaudum) [inline src/crate/crash/command.py]: typo: system

**Comment by mfussenegger** [comment]
would it make sense to extract the sysinfo stuff into it's own module?
I think at some point we could make command extensible via plugins and then it would be good to not have them in the CrateCmd class

File: src/crate/crash/command.py [hunk 5]
```diff
@@ -322,6 +329,74 @@ def _connect(self, server):
...
+                                os_info['jvm']['version'],
+                                os_info['jvm']['vm_vendor'],
+                                os_info['jvm']['vm_name'],
+                                os_info['jvm']['vm_version']) as jvm_info
+                          from sys.nodes
+                          order by os['uptime'] desc """)
+        self.pprint(self.cursor.fetchall(), [c[0] for c in self.cursor.description])
+
+    def _quantity_sysinfo(self):
+            self.rows = []
```
Comment (chaudum) [inline src/crate/crash/command.py]: indentation!

File: src/crate/crash/command.py [hunk 6]
```diff
@@ -322,6 +329,74 @@ def _connect(self, server):
             self.logger.info('CONNECT OK')
             self._check()
 
+    def sysinfo(self):
+        prev_format = self.output_writer.output_format
+        self._switch_format('mixed')
+        self._sysinfo()
```
Comment (chaudum) [inline src/crate/crash/command.py]: make sure the output format is reset in all error cases

File: src/crate/crash/command.py [hunk 7]
```diff
@@ -322,6 +329,74 @@ def _connect(self, server):
...
+                          format('java version \"%s\" %s %s (build %s)',
+                                os_info['jvm']['version'],
+                                os_info['jvm']['vm_vendor'],
+                                os_info['jvm']['vm_name'],
+                                os_info['jvm']['vm_version']) as jvm_info
+                          from sys.nodes
+                          order by os['uptime'] desc """)
+        self.pprint(self.cursor.fetchall(), [c[0] for c in self.cursor.description])
+
+    def _quantity_sysinfo(self):
```
Comment (chaudum) [inline src/crate/crash/command.py]: function naming?

File: src/crate/crash/command.py [hunk 8]
```diff
@@ -322,6 +329,74 @@ def _connect(self, server):
...
+            self.rows = []
+            self.cols = []
+            self._num_nodes()
+            self._num_tables()
+            self._num_records_and_shards()
+            self.pprint([self.rows], self.cols)
```
Comment (chaudum) [inline src/crate/crash/command.py]: combine `_num_nodes()`, `self._num_tables()` and `self._num_records_and_shards()` into a single function call

File: src/crate/crash/command.py [hunk 9]
```diff
@@ -322,6 +329,74 @@ def _connect(self, server):
...
+        """ print system and cluster info """
+        if self.connection.lowest_server_version >= SYSINFO_MIN_VERSION:
+            self._quantity_sysinfo()
+            self._nodes_info()
+        else:
+            tmpl = 'Crate {version} does not support the cluster "sysinfo" command'
+            self.logger.warn(tmpl.format(version=self.connection.lowest_server_version))
+
+    def _nodes_info(self):
+        self._execute(""" select version['number'] as crate_version,
```
Comment (chaudum) [inline src/crate/crash/command.py]: make the query strings constants

**Comment by chaudum** [comment]
I agree with @mfussenegger 
having a separate module would make the code much cleaner!

File: src/crate/crash/command.py [hunk 10]
```diff
@@ -201,8 +208,10 @@ def __init__(self,
             'connect': self._connect,
             'dt': self._show_tables,
             'check': self._check,
+            'sysinfo': self._sys_info,
         }
         self.logger = ColorPrinter(is_tty)
+        self.sysinfo = SysInfo(connection, error_trace, is_tty)
```
Comment (chaudum) [inline src/crate/crash/command.py]:
```python
self.sysinfo = SysInfoCommand(self, *kwargs)
```

File: src/crate/crash/sysinfo.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            self.cmd.logger.info("For debugging purposes you can send above listed information to support@crate.io")
+        else:
+            tmpl = 'Crate {version} does not support the cluster "sysinfo" command'
+            self.cmd.logger.warn(tmpl.format(version=self.cmd.connection.lowest_server_version))
+
+    def _sys_info(self):
+        self.result = []
+        success = self._cluster_info()
+        success &= self._nodes_info()
+        return (success, self.result)
```
Comment (chaudum) [inline src/crate/crash/sysinfo.py]: why returning an instance variable?

File: src/crate/crash/sysinfo.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+            if success is False:
+                return success
+            rows.extend(self.cmd.cursor.fetchall()[0])
+            cols.extend([c[0] for c in self.cmd.cursor.description])
+        self.result.append(Result([rows], cols))
+        return True
+
+    def _nodes_info(self):
+        success = self.cmd._execute(SysInfoCommand.NODES_INFO[0])
+        if success:
```
Comment (chaudum) [inline src/crate/crash/sysinfo.py]: ``py
if self.cmd._execute(SysInfoCommand.NODES_INFO[0]):
    self.result.append(...)
```

File: CHANGES.txt [hunk 13]
```diff
@@ -5,6 +5,9 @@ Changes for crash
 Unreleased
 ==========
 
+ - Added the ``sysinfo`` command which queries the ``sys`` tables for
+   system and cluster information.
```
Comment (chaudum) [inline CHANGES.txt]: maybe add: this requires Crate `>= 0.54.0`

**Comment by chaudum** [comment]
otherwise ack

**Pull request merged**